### PR TITLE
Linux: use libc in triplet instead of distro

### PR DESCRIPTION
**Pull request by haampie**
Spack over-specializes the triplet on Linux distro (and version), which makes buildcaches on Linux rather difficult to use across distros. This PR is an attempt to move it towards more generic triplet a la linux-gnu and linux-musl.

The way it's currently done: the host os is detected in terms of `<glibc/musl><ver>` instead of `<distro name><ver>` when possible, the os compat rules are relaxed with a two way `os_compatible` mapping between distro and libc (for example `ubuntu18.04` / `glibc2.27`) and one way `os_compatible` rules encoding known ABI compatibility of libc, and minimization on mismatches is removed as a criterion to allow reuse by default.

Host libc is detected through the first working method in order:

1. `confstr(3)` with `_CS_GNU_LIBC_VERSION`, which is a cheap way to detect glibc >= 2.3.2
2. Parsing of PT_INTERP of currently running Python, and checking if it matches `ld-musl-*`. If so, run the dynamic linker with `--version` flag, which is a reasonably cheap way to detect musl libc, although it requires a subprocess
3. Parse `ldd --version` output, which depends on PATH (not great, but OK)

If none work, the original `distro` implementation kicks in.

(Initially I also had an implementation to detect glibc based on version symbol definitions, but then I realized the dynamic linker obtained through PT_INTERP does not define the latest `GLIBC_*` symbol, so you'd have to locate libc.so, which is also sensitive to environment variables, just like `ldd --version`, and the latter is simpler anyways :(. Also tried `platform.libc_ver()` but it's a real mess and plain wrong on older versions of Python)

---

Regarding the concretizer: I think it makes sense to drop minimization on OS mismatches, since it blocks reuse. Especially on rolling releases, it is very convenient to be able to mix nodes that were once linked again `glibc2.20`, `glibc2.21`, ... since system glibc could be bumped every so many months.

---

The usefulness/uselessness of this PR now depends on `%compiler` bits in specs. This PR is still kind of useless when a strict compiler mismatch rule is imposed (which is the case right now), but that's not easily addressed, because there are two use cases that warrant different rules: 

1. If you need to locally build on top of prebuilt binaries, you likely need strict compiler version matching, because of lack of guarantees on forward compatibility of file formats of Fortran / C++ modules, LTO data, pre-compiled headers etc (note: I didn't verify if/how incompatible these formats are)
2. If you don't even have a compiler installed, and simply want to use `spack install` like it's `apt install` to pull in some binaries, the compiler compat rules are overly strict (for example we could relax them by parsing symbol versions of `lib{stdc++,gfortran}` and then turn those into `%gcc@x.y:` constraints)

I guess right now the PR is only useful as long as the same compilers are used on different distros; users no longer have to contribute `os_compatible` rules to Spack once their os is defined in terms of glibc/musl.

### CODE REVIEW

**Comment by haampie** [comment]
There are only a handful packages that define conditionson the OS:

```python
% rg --pcre2 '\bos=(?!(ventura|catalina|bigsur|monterey|mojave|highsierra))' -g '*.py'                                                       (fix/distro-to-libc)spack
packages/berkeley-db/package.py
90:            spec.satisfies("os=centos7") or spec.satisfies("os=rhel8")

packages/relion/package.py
97:    patch("0002-Simple-patch-to-fix-intel-mkl-linking.patch", when="@:3.1.1 os=ubuntu18.04")

packages/openssl/package.py
394:        if self.spec.satisfies("%nvhpc os=centos7"):

packages/gcc/package.py
342:    conflicts("~binutils", when="@7: os=rhel6", msg="New GCC cannot use system assembler on RHEL6")
346:        "~binutils", when="@10: os=rhel7", msg="gcc: Add +binutils - preinstalled as might be old"
349:        "~binutils", when="@10: os=rhel8", msg="gcc: Add +binutils - preinstalled as might be old"

packages/process-in-process/package.py
32:    conflicts("%gcc@:3", when="os=centos7")
33:    conflicts("%gcc@5:", when="os=centos7")
34:    conflicts("%gcc@:3", when="os=rhel7")
35:    conflicts("%gcc@5:", when="os=rhel7")
36:    conflicts("%gcc@:7", when="os=centos8")
37:    conflicts("%gcc@9:", when="os=centos8")
38:    conflicts("%gcc@:7", when="os=rhel8")
39:    conflicts("%gcc@9:", when="os=rhel8")
57:        when="@2 os=centos7",
64:        when="@2 os=rhel7",
72:        when="@2 os=centos8",
79:        when="@2 os=rhel8",
90:            when="@2 os=centos7",
97:            when="@2 os=rhel7",
105:            when="@2 os=centos8",
112:            when="@2 os=rhel8",
123:            when="@3 os=centos7",
130:            when="@3 os=rhel7",
138:            when="@3 os=centos8",
145:            when="@3 os=rhel8",
156:                when="@3 os=centos7",
163:                when="@3 os=rhel7",
171:                when="@3 os=centos8",
178:                when="@3 os=rhel8",

packages/glib/package.py
143:    patch("old-kernels.patch", when="@2.56.0:2.56.1 os=rhel6")
144:    patch("old-kernels.patch", when="@2.56.0:2.56.1 os=centos6")
145:    patch("old-kernels.patch", when="@2.56.0:2.56.1 os=scientific6")

packages/intel-tbb/package.py
240:            spec.satisfies("%gcc@4.8.0: os=rhel6")
241:            or spec.satisfies("%gcc@4.8.0: os=centos6")
242:            or spec.satisfies("%gcc@4.8.0: os=scientific6")
```

some of these are also incorrect "system assembler is too old" even though you can install a new binutils on that distro.

**Comment by haampie** [comment]
The usefulness/uselessness of this PR now depends on `%compiler` bits in specs

**Comment by haampie** [comment]
@kwryankrattiger can you somehow nuke the relevant S3 bucket for this PR? The PR was opened a long time ago, and now when running Gitlab CI part of the binaries may have disappeared from the PR cache?

**Comment by kwryankrattiger** [comment]
Closing and reopening should have done it, but I can double check.

**Comment by haampie** [comment]
@kwryankrattiger https://gitlab.spack.io/spack/spack/-/jobs/7898568 for example
```
==> Error: No binary for glproto-1.4.17-v7vgkpmetautpwr7j76dfmw4jwhrcdim found when cache-only specified
```

I wonder if spec.json files exist but tarballs are gone?

**Comment by scottwittenburg** [comment]
I re-ran this pipeline after the failures of 2 days ago, since the buildcache indices were updated by that failed pipeline, and any weird mismatches should have been fixed.  But jobs still failed, so I looked a bit more.  The jobs that failed seemed to have in common that the deps they couldn't install were not listed as jobs they `need`-ed, either directly or transitively.

Can you think of anything about this PR that would have an effect on how we compute dependencies for pipeline job scheduling?  E.g. in the `ml-linux-x86_64-cpu` stack, `py-tensorflow` failed when it tried to install `py-libclang`, and while there were several `py-libclang` built in the pipeline, no job in the generated pipeline yaml actually `need`s it.  

Or is there something obviously wrong about how I'm traversing the environment specs, [here](https://github.com/spack/spack/blob/develop/lib/spack/spack/ci.py#L310-L325), to build the pipeline graph?  I don't think I'm seeing this weird behavior anywhere else at the moment, but maybe the PR uncovered a bug in pipeline generation that I can fix.

**Comment by haampie** [comment]
> Can you think of anything about this PR that would have an effect on how we compute dependencies for pipeline job scheduling?

Not immediately :thinking: 

> Or is t[here](https://github.com/spack/spack/blob/develop/lib/spack/spack/ci.py#L310-L325) something obviously wrong about how I'm traversing the environment specs, here, to build the pipeline graph?

You should probably use `spack.traverse.traverse_edges(spec_list, root=False, deptype=all, cover="edges")` so you visit each edge exactly once (spec_list is environment roots?) but that's not a bug, just a performance improvement (well, a performance bug :p). Also, it's not a great idea to use dicts everywhere where you can use (named) tuples... that's tons of useless allocations, hash function calls, and type erasure.

**Comment by haampie** [comment]
@scottwittenburg it looks like you're populating `needs` only with direct deps:

https://github.com/spack/spack/blob/0b4631a7741df403510281b304d7bd6b169beebc/lib/spack/spack/ci.py#L1043

That doesn't seem to take into account that direct deps can be pruned, even if their deps are being rebuilt (e.g. someone just deleted a random spec.json/tarball or it's not available for download, so the buildcache index is in an inconsistent state where a required dep is `installed: false`).

I don't understand why pruning is not done on the graph: if a node is removed, its in and out edges should be removed, and new edges between its dependencies and dependents should be added back (the latter is not done rn)

File: lib/spack/spack/solver/asp.py [hunk 1]
```diff
@@ -1847,7 +1847,7 @@ def os_defaults(self, specs):
         # Consider any OS's mentioned on the command line. We need this to
         # cross-concretize in CI, and for some tests.
         # TODO: OS should really be more than just a label -- rework this.
-        for spec in specs:
+        for spec in spack.traverse.traverse_nodes(specs):
```
Comment (alalazo) [inline lib/spack/spack/solver/asp.py:1850]: Is this specific to this PR, or just an optimization to avoid multiple traversals of input specs?

File: lib/spack/spack/test/relocate.py [hunk 2]
```diff
@@ -172,7 +165,7 @@ def test_normalize_relative_paths(start_path, relative_paths, expected):
 
 
 @pytest.mark.requires_executables("patchelf", "file", "gcc")
-@skip_unless_linux
+@pytest.mark.skipif(sys.platform != "linux", reason="only tested on linux")
```
Comment (alalazo) [inline lib/spack/spack/test/relocate.py:168]: The changes in this file are also orthogonal to the PR.

File: lib/spack/spack/test/util/elf.py [hunk 3]
```diff
@@ -17,17 +18,8 @@
...
-def skip_unless_linux(f):
-    return pytest.mark.skipif(
-        str(spack.platforms.real_host()) != "linux",
-        reason="implementation currently requires linux",
-    )(f)
-
-
 @pytest.mark.requires_executables("gcc")
-@skip_unless_linux
+@pytest.mark.skipif(sys.platform != "linux", reason="only tested on linux")
```
Comment (alalazo) [inline lib/spack/spack/test/util/elf.py:22]: Same. We can consolidate the fixtures and provide `linux_only`, since we now have `windows_only`

**Review by alalazo — COMMENTED** [review]
This is just a first series of thoughts, after playing with the PR for a bit. 

---

On first install, I had to run:
```console
spack compiler find
```
again to have Spack concretizing. This is due to the different value that the attribute `operating_system` has for compilers in this PR. Is this expected?

---

We have a dummy `libc` package, which I think is mainly used to mark external providers for `rpc` and `iconv`. I wonder how far are we from removing the `os` attribute conditionally, and add a dependency on the detected `libc` instead. That would ideally imply splicing when we reuse a buildcache made with a compatible glibc.[^1]

[^1]: That probably needs a more extensive discussion on how to model ABI compatibility, and on whether to use "virtual" packages to model runtime libraries.
---

I'm not opposed to this transient change of having `os=<libc-version>` but I'm trying to think if we can avoid multiple breaking changes on users. I guess in the end we want to get rid of the `os=...` label and have a libc node in the DAG, starting very likely with just an "external" node. What would change / be inconvenient for users:
- When moving first to `os=<libc-version>`
- When moving later to `libc` as a node

? Did we ever discuss or planned this?

---

The detection of libc seems gold, I think we'll reuse it no matter what would be the first implementation.

**Comment by alalazo** [comment]
Other considerations on top of this. 

If we drop `os=` completely and add a leaf node for `libc` we might want to remove the compiler attribute from the leaf node too, to avoid issues due to compiler matching in concretization[^1].

Can we reliably disregard `target` compatibility when dealing with `libc`? Meaning, do we have use cases where a `libc` is compiled to support only targets > X and we have to track that?

If/when we go with a leaf node, the detection logic should be moved to a virtual `package`, and generate externals like:
```yaml
packages:
  # libc is a virtual package
  libc:
  - spec: gibc@2.X
    prefix: ...
```

[^1]: We can also special case `libc`, but I'd avoid ad-hoc hacks if possible

**Comment by haampie** [comment]
Closing this for the following reasons:

1. Instead of a `GNU/linux` bit in a triplet, we're going to add `^glibc@x.y` as a dependency. This makes `conflicts` etc possible too.
2. We shouldn't figure out the libc flavor and version from the Python interpreter, but rather from the compiler. Inferring it from the Python interpreter is wrong if you would use a statically compiled Python from elsewhere.

Follow-up is in #43164
